### ISSUE TITLE
[Http Instrumentation] Make it easier to scrub sensitive URI details from output

### ISSUE THREAD

**Issue by stephenjust** [open-telemetry/opentelemetry-dotnet-contrib#1791]
# Feature Request

Before opening a feature request against this repo, consider whether the feature
should/could be implemented in the [other OpenTelemetry client
libraries](https://github.com/open-telemetry/). If so, please [open an issue on
opentelemetry-specification](https://github.com/open-telemetry/opentelemetry-specification/issues/new)
first.

**Is your feature request related to a problem?**

If so, provide a concise description of the problem.

> The HTTP instrumentation package currently writes out the full request path for most requests, the exception being requests that contain Uri.UserInfo. This is not sufficient for many resource types. For example, the Azure Storage SDK's requests show up in logs with a SAS token appended. 
>
> For example:
> `https://mystorage.blob.core.windows.net/container/myfile.txt?sv=2019-07-07&sr=c&sig=placeholder&se=2021-09-08T22%3A03%3A23Z&sp=r&api-version=2019-07-07&`
>
> I can also imagine other scenarios where the URI might otherwise contain sensitive information that cannot be logged.

**Describe the solution you'd like:**

What do you want to happen instead? What is the expected behavior?

> It would be helpful if OpenTelemetry had a configurable high-performance URL scrubber for anywhere URLs could be logged,
> to make it as easy as possible for developers to sanitize their log output. This would support (minimally) blocking out any specified query string parameters.

**Describe alternatives you've considered.**

Which alternative solutions or features have you considered?

> In our project today, we configure an enricher to run a Regex replace on all URLs we log, but it feels like there is room for improvement on the default behavior.

## Additional Context

Add any other context about the feature request here.

> This is definitely in the "nice to have" category. There are ways to solve the problem with the current shipped libraries.

**Comment by Kielek** [comment]
Closing. Latest released brings `OTEL_DOTNET_EXPERIMENTAL_HTTPCLIENT_DISABLE_URL_QUERY_REDACTION` to redact all data. It can be disabled by mentioned environmental variable.

**Comment by swythan** [comment]
> Closing. Latest released brings `OTEL_DOTNET_EXPERIMENTAL_HTTPCLIENT_DISABLE_URL_QUERY_REDACTION` to redact all data. It can be disabled by mentioned environmental variable.

That isn't equivalent. 

1. It defaults the other way. That setting _prevents_ the reaction of all quiet parameters that was just defaulted to being enabled. (I'll leave the heated discussions to the other threads on that one.)
2. What I want is to be able to redact _only_ the sensitive query parameters. The original request says _configurable_ so I don't think I'm alone in that.

**Comment by Kielek** [comment]
@swythan, reopening per your request.

**Comment by acu-kevinb** [comment]
I just raised a related item. It's actually possible to achieve this using a the pre-release OpenTelemetry.Extensions.Enrichment.AspNetCore, and injecting a IHttpContextAccessor into your implementation of AspNetCoreTraceEnricher. You can then add metadata to specific endpoints, and use that data to configure per-endpoint trace transformation.

https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/3933

Edit:
Sorry, I misread this and thought it was related to the aspnetcore instrumentation and not httpclient.

**Comment by PetarJerinic** [comment]
I volunteer to take this one. Semconv 1.44.0 (there were changes as far as I can see from the time of the discussion) [UrlAttributes](https://opentelemetry.io/docs/specs/semconv/registry/attributes/url/#url-attributes) has this under development badge:
> Instrumentation MAY provide a way to override this list via declarative configuration. If so, it SHOULD use the sensitive_query_parameters property (an array of case-sensitive strings with minimum items 0)

I suggest the approach to extend `Instrumentation.Http` with the following:

-  `UrlQueryRedactionMode` containing AllValues and SensitiveParameters options. Default would be AllValues to avoid breaking current behavior. It will be added to `HttpClientTraceInstrumentationOptions`
- adding `SensitiveQueryParameters` collection to `HttpClientTraceInstrumentationOptions`. It will be seeded with 5 semconv keys and overridable.

I'll create a PR in the next few days unless there are different directional suggestions or objections from maintainers.

**Comment by martincostello** [comment]
> via declarative configuration

Note that [declarative configuration](https://opentelemetry.io/docs/languages/sdk-configuration/declarative-configuration/) is a specific concept that we do not yet implement anywhere in the .NET SDK.

It doesn't necessarily mean we won't take changes to implement this based on what's in SemConv, but it shouldn't be read as just a generic configuration option to implement.

As the HTTP instrumentation package is stable, any implementation of non-stable features will likely need to be configurable only through experimental environment variables, rather than first-class API surface.

**Comment by PetarJerinic** [comment]
Agreed about declarative configuration, due to it (currently) not being implemented in .NET my proposal went in direction of extending public API.

Good point it is non-stable feature, I’d implement it with current conventions in mind and go toward having single experimental env variable (simplifying my proposal above and also more in line with semconv wording): `OTEL_DOTNET_EXPERIMENTAL_HTTPCLIENT_SENSITIVE_QUERY_PARAMETERS` (comma-separated values).
